src/diag/map_editor.cpp

```diff
@@ -5,149 +5,96 @@
-            return 0;
-        } else {
-            const char* name = map_id_to_name(map_idx);
-            if (name == nullptr) {
-                ESP_LOGE("MAP_EDITOR_R", "map name is null!?");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-            if (EEPROM::read_nvs_map_data(name, b, nullptr, SHIFT_MAP_SIZE)) {
-                // OK, copy the pointer
-                *buffer = b;
-                *dest_size = SHIFT_MAP_SIZE*sizeof(int16_t);
-                return 0;
-            } else {
-                // De-allocate
-                ESP_LOGE("MAP_EDITOR_R", "read_nvs_map_data failed!");
-                delete b;
-                return NRC_GENERAL_REJECT;
-            }
-        }
+uint8_t MapEditor::read_map_data(uint8_t map_id, bool is_default, uint16_t *dest_size, int16_t** buffer) {
+    CHECK_MAP(map_id)
+    // Map valid
+    int size = ptr->get_map_element_count();
+    int16_t* b = (int16_t*)heap_caps_malloc(size*sizeof(int16_t), MALLOC_CAP_SPIRAM);
+    if (b == nullptr) {
+        ESP_LOGE("MAP_EDITOR_R", "Could not allocate read array!");
+        return NRC_GENERAL_REJECT;
+    }
+    const int16_t* data;
+    if (is_default) {
+        data = ptr->get_default_map_data();
     } else {
-        return NRC_SUB_FUNC_NOT_SUPPORTED_INVALID_FORMAT;
+        data = (const int16_t*)ptr->get_current_map_data();
     }
+    memcpy(b, data, size*sizeof(int16_t));
+    *buffer = b;
+    *dest_size = size*sizeof(int16_t);
+    return 0;
 }
     
 uint8_t MapEditor::write_map_data(uint8_t map_id, uint16_t dest_size, int16_t* buffer) {
-    if (map_id >= 0x01 && map_id <= 0x0C) {
-        const char* name = map_id_to_name(map_id);
-        if (name == nullptr) {
-            ESP_LOGE("MAP_EDITOR_W", "map name is null!?");
-            return NRC_GENERAL_REJECT;
-        }
-        if (dest_size != SHIFT_MAP_SIZE*sizeof(int16_t)) {
-            ESP_LOGE("MAP_EDITOR_W", "Buffer has %d bytes. Shift map needs %d bytes", dest_size, sizeof(int16_t)*SHIFT_MAP_SIZE);
-            return NRC_GENERAL_REJECT;
-        }
-        if (EEPROM::write_nvs_map_data(name, buffer, SHIFT_MAP_SIZE)) {
-            spkr.send_note(1500, 300, 310);
+    CHECK_MAP(map_id)
+    if (ptr->replace_map_content(buffer, dest_size)) {
+        return 0;
+    } else {
+        return NRC_GENERAL_REJECT;
+    }
+}
+
+uint8_t MapEditor::burn_to_eeprom(uint8_t map_id) {
+    CHECK_MAP(map_id)
+    if (ptr->save_to_eeprom()) {
+        return 0;
+    } else {
+        return NRC_GENERAL_REJECT;
+    }
+}
 
-            return 0;
-        } else {
-            ESP_LOGE("MAP_EDITOR_W", "write_nvs_map_data failed!");
-            return NRC_GENERAL_REJECT;
-        }
-    }  else {
-        return NRC_SUB_FUNC_NOT_SUPPORTED_INVALID_FORMAT;
+uint8_t MapEditor::reset_to_program_default(uint8_t map_id) {
+    CHECK_MAP(map_id)
+    if (ptr->reload_from_eeprom() && ptr->save_to_eeprom()) {
```

Comment:
Dangerous, since you don't know, in which order the compiler will evaluate the two functions. Consider splitting it into two if-clauses.

Reply:
https://eel.is/c++draft/expr.log.and
 
> The && operator groups left-to-right[.](https://eel.is/c++draft/expr.log.and#1.sentence-1) The operands are both
> contextually converted to bool. The result is true if both operands are true and false otherwise. Unlike &, && 
> guarantees left-to-right evaluation: the second operand is not evaluated if the first operand is false.

Reply:
Agreed, always thought lazy evaluation is up to the compiler, but it seems I've been mistaken about that.